File: Find_first_and_Last_position_in_sorted_array.py
#Time Complexity: O(log n)
#Space Complexity: O(1)

class Solution(object):
        
        #Binary search in the first half of the array
        def binarySearchFirst(nums, target):
            low = 0
            high = len(nums)-1
            
            while (low <= high):
                mid = low + (high - low)//2
                if target == nums[mid]: #checking if target is the value at mid
                    if (mid == 0) or (nums[mid] > nums[mid-1]): #since sorted array, true here means start index of target will not exist before mid
                        return mid  #true condition stated mid is the start position
                    else:
                        high = mid - 1
                elif target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            return -1
                    
        
        #search the second half for last position          
        def binarySearchLast(nums, target):
            low = 0
            high = len(nums)-1
            
            while (low <= high):
                mid = low + (high - low)//2
                if target == nums[mid]: #checking if target is the value at mid
                    if (mid == len(nums)-1) or (nums[mid] < nums[mid+1]): #since sorted array, true here means start index of target will not exist before mid
                        return mid  #true condition stated mid is the start position
                    else:
                        low = mid + 1
                elif target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1
            return -1
        
        #search for the range 
        def searchRange(self, nums, target):
            if (nums == None) or (len(nums) == 0):
                return (-1,-1)
            if (target < nums[0] or target > nums[len(nums) - 1]):
                return (-1,-1)
            
            firstIndex = Solution.binarySearchFirst(nums, target)
            lastIndex = Solution.binarySearchLast(nums, target)
            
            return (firstIndex, lastIndex)

File: FindFirstAndLastPositionInsortedArray.java
//Time Complexity :O(logN) +O(logN) = O(logN)
//Space Complexity :O(1)
//https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if(nums==null || nums.length ==0) return new int[]{-1,-1};
        int low=0;
        int high=nums.length-1;
        if(target<nums[low] || nums[high]<target) return new int[]{-1,-1};
        int first = leftSearch(nums,low,high,target);
        if(first == -1) return new int[]{-1,-1};
        int last = rightSearch(nums,first,high,target);
        
        return new int[]{first,last};
    }
    public int leftSearch(int [] nums,int low,int high,int target){
        while(low<=high){
            int mid = low+ (high-low)/2;
            if(nums[mid]==target){
                if(mid==0 || target!=nums[mid-1]) {
                       return mid;
                   }
                   high =mid-1;
                }else if(nums[mid] <target){
                low = mid+1;
            }else{
                high=mid-1;
            }
        }
        return -1;
    }
    public int rightSearch(int [] nums,int low,int high,int target){
        while(low<=high){
            int mid = low+ (high-low)/2;
            if(nums[mid]==target){
                if(mid==nums.length-1 || target!=nums[mid+1]) {
                       return mid;
                   }
                   low =mid+1;
                }else if(nums[mid] <target){
                low = mid+1;
            }else{
                high=mid-1;
            }
        }
        return -1;
    }
}
